### DISCUSSION TITLE
Twice-forwarded args and kwargs

### DISCUSSION THREAD

**Discussion by dpmccabe** [microsoft/pyright#7975] [Q&A] [answered]
I have a function that is designed to retry an arbitrary function call some number of times:

```py
P = ParamSpec("P")
R = TypeVar("R")

def maybe_retry(
    func: Callable[P, R],
    retryable_exceptions: tuple[Type[Exception], ...] = tuple([Exception]),
    max_attempts: int = 0,
    *args: P.args,
    **kwargs: P.kwargs,
) -> R:
    if max_attempts == 0:
        return func(*args, **kwargs) # etc.
```

I have another function that wraps API requests and handles responses:

```py
def call_api(func: Callable, *args: object, **kwargs: object) -> Any:
    res = maybe_retry(
        func,
        retryable_exceptions=(requests.ConnectionError, requests.ConnectTimeout),
        *args,
        **kwargs,
    ) # etc.
```

While this code works, Pyright gives me this error:

```
Argument of type "object" cannot be assigned to parameter "max_attempts" of type "int" in function "maybe_retry"
    "object" is incompatible with "int" (reportArgumentType)
```

Basically, any `maybe_retry` keyword argument that I don't specify when I call it results in such an error. Is there any way to resolve this? I'm open to refactoring and dropping the option for positional args if that would make the typing easier.

**Comment by erictraut** [comment, marked as answer]
Pyright is doing the right thing here. For comparison, mypy emits the same error.

One workaround is to replace the `object` annotations with `Any` in the `call_api` signature.

If you want to retain more type safety, you can do the following:

```python
def call_api(
    func: Callable[P, R], max_attempts: int = 0, *args: P.args, **kwargs: P.kwargs
) -> R:
    return maybe_retry(
        func,
        retryable_exceptions=(requests.ConnectionError, requests.ConnectTimeout),
        max_attempts=max_attempts,
        *args,
        **kwargs,
    )
```

**Reply by dpmccabe** [reply]
`Any` is a perfectly fine solution for me this time. Thanks for the quick response.
